src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/ResourceScope.java

```diff
@@ -0,0 +1,289 @@
+ * as such resources associated with this shared resource scopes can be accessed by multiple threads.
+ * This might be useful when multiple threads need to access the same resource concurrently (e.g. in the case of parallel processing).
+ * For instance, a client might obtain a {@link Spliterator} from a shared segment, which can then be used to slice the
+ * segment and allow multiple threads to work in parallel on disjoint segment slices. The following code can be used to sum
+ * all int values in a memory segment in parallel:
+ *
+ * <blockquote><pre>{@code
+SequenceLayout SEQUENCE_LAYOUT = MemoryLayout.sequenceLayout(1024, MemoryLayouts.JAVA_INT);
+try (ResourceScope scope = ResourceScope.newSharedScope()) {
+    MemorySegment segment = MemorySegment.allocateNative(SEQUENCE_LAYOUT, scope);
+    VarHandle VH_int = SEQUENCE_LAYOUT.elementLayout().varHandle(int.class);
+    int sum = StreamSupport.stream(segment.spliterator(SEQUENCE_LAYOUT), true)
+        .mapToInt(s -> (int)VH_int.get(s.address()))
+        .sum();
+}
+ * }</pre></blockquote>
+ *
+ * <p>
+ * Explicit shared resource scopes, while powerful, must be used with caution: if one or more threads accesses
+ * a resource associated with a shared scope while the scope is being closed from another thread, an exception might occur on both
+ * the accessing and the closing threads. Clients should refrain from attempting to close a shared resource scope repeatedly
+ * (e.g. keep calling {@link #close()} until no exception is thrown). Instead, clients of shared resource scopes
+ * should always ensure that proper synchronization mechanisms (e.g. using resource scope handles, see below) are put in place
+ * so that threads closing shared resource scopes can never race against threads accessing resources managed by same scopes.
+ *
+ * <h2>Resource scope handles</h2>
+ *
+ * Resource scopes can be made <em>non-closeable</em> by acquiring one or more resource scope <em>handles</em> (see
+ * {@link #acquire()}. A resource scope handle can be used to make sure that resources associated with a given resource scope
+ * (either explicit or implicit) cannot be released for a certain period of time - e.g. during a critical region of code
+ * involving one or more resources associated with the scope. For instance, an explicit resource scope can only be closed
+ * <em>after</em> all the handles acquired against that scope have been closed (see {@link Handle#close()}).
+ * This can be useful when clients need to perform a critical operation on a memory segment, during which they have
+ * to ensure that the segment will not be released; this can be done as follows:
+ *
+ * <blockquote><pre>{@code
+MemorySegment segment = ...
+ResourceScope.Handle segmentHandle = segment.scope().acquire()
+try {
+   <critical operation on segment>
+} finally {
+   segment.scope().release(segmentHandle);
+}
```

Comment:
I do like the symmetry in this example, but just to add an alternative idiom:
    `segmentHandle.scope().release(segmentHandle)`
, which guarantees to avoid release throwing and IAE

Reply:
I see what you mean - I don't think I want to encourage that style too much by giving it prominence in the javadoc. It's true you can go back to the scope from the handle, so that you are guaranteed to release the right thing, but I think that should be unnecessary in most idiomatic uses.